### PR TITLE
Create data stream aliases from template

### PR DESCRIPTION
**Pull request by danhermann**
Allows the creation of data stream aliases from a data stream template.

Relates to #66163

### CODE REVIEW

**Comment by danhermann** [comment]
@elasticmachine update branch

**Comment by danhermann** [comment]
@elasticmachine update branch

**Comment by danhermann** [comment]
@elasticmachine update branch

**Comment by danhermann** [comment]
@martijnvg, I have data stream aliases being created from templates in this PR. I did it using only the `AliasMetadata` class without using the `o.e.cluster.metadata.DataStreamAlias` class which works as long as we're willing to overload or rename some of the members of `AliasMetadata` to refer to indices or data streams. If you think this is a good approach, I'll go ahead and add some tests and get it merged.

File: server/src/main/java/org/elasticsearch/cluster/metadata/AliasMetadata.java [hunk 1]
```diff
@@ -51,8 +52,11 @@
     @Nullable
     private final Boolean isHidden;
 
+    @Nullable
+    private final Boolean isDataStream;
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/cluster/metadata/AliasMetadata.java:56]: I don't think this field is needed? I also don't see that is currently is used in this pr.
In `MetadataCreateDataStreamService`, data stream aliases can be created without this information? The fact that a `DataStreamTemplate` is defined should be sufficient?

File: server/src/main/java/org/elasticsearch/cluster/metadata/ComposableIndexTemplate.java [hunk 2]
```diff
@@ -256,35 +259,63 @@ public String toString() {
...
+                    aliasMap.put(alias.alias(), alias);
+                }
+                return aliasMap;
+            }, ALIASES);
         }
 
         private final boolean hidden;
 
+        @Nullable
+        private final Map<String, AliasMetadata> aliases;
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/cluster/metadata/ComposableIndexTemplate.java:293]: I think we can reuse the alias definitions in `ComposableIndexTemplate` (defined inside the `Template` class)?
If a composable index template has a `DataStreamTemplate` and aliases definitions are present then data stream aliases are created. If just alias definitions are present in composable index template then indices aliases are created.

**Review by martijnvg — COMMENTED** [review]
Thanks @danhermann, I think this pr is in the right direction, but I think we maybe can simplify even more (see 2 comments).

**Review by martijnvg — COMMENTED** [review]
Looks good. I left a few more comments.

**Comment by danhermann** [comment]
@elasticmachine update branch

**Comment by danhermann** [comment]
@elasticmachine run elasticsearch-ci/part-2

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java [hunk 3]
```diff
@@ -519,14 +520,27 @@ private ClusterState applyCreateIndexRequestWithV2Template(final ClusterState cu
...
+        return applyCreateIndexWithTemporaryService(
+            currentState,
+            request,
+            silent,
+            null,
+            tmpImd,
+            mappings,
+            indexService -> resolveAndValidateAliases(
+                request.index(),
+                isDataStream ? Set.of() : request.aliases(),
```
[the remark this answers is not in the document]
Comment (danhermann) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java:533]: Thanks, @martijnvg. I've made those changes.

File: server/src/test/java/org/elasticsearch/cluster/metadata/MetadataCreateDataStreamServiceTests.java [hunk 4]
```diff
@@ -66,6 +69,41 @@ public void testCreateDataStream() throws Exception {
...
     }
 
+    public void testCreateDataStreamWithAliasFromTemplate() throws Exception {
+        final MetadataCreateIndexService metadataCreateIndexService = getMetadataCreateIndexService();
+        final String dataStreamName = "my-data-stream";
+        final int aliasCount = randomIntBetween(0, 3);
+        Map<String, AliasMetadata> aliases = new HashMap<>(aliasCount);
+        for (int k = 0; k < aliasCount; k++) {
+            final String aliasName = randomAlphaOfLength(6);
+            aliases.put(aliasName, AliasMetadata.newAliasMetadataBuilder(aliasName).build());
```
Comment (martijnvg) [inline server/src/test/java/org/elasticsearch/cluster/metadata/MetadataCreateDataStreamServiceTests.java:79]: maybe also test the case here with aliases with write data streams and filter? (Either by randomly adding and asserting it or a new test)

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateDataStreamService.java [hunk 5]
```diff
@@ -222,6 +221,13 @@ static ClusterState createDataStream(MetadataCreateIndexService metadataCreateIn
         logger.info("adding data stream [{}] with write index [{}] and backing indices [{}]", dataStreamName,
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateDataStreamService.java:221]: maybe enhance this log line with the fact that aliases were created as well (if template has aliases)?

**Review by martijnvg — APPROVED** [review]
Left 2 small comments, LGTM otherwise 👍.

**Comment by danhermann** [comment]
Thanks, @martijnvg. I made those changes.

**Pull request merged**